[ThinLTO][Cache] Allow ThinLTO to use RemoteCachingService as cache

cachemeifyoucan opened the pull request:
Add implementation to allow ThinLTO to use RemoteCachingService as cache, it can be enabled by either setting the cache dir to URL path like "grpc:/$SERVICE_SOCKET_PATH" or setting the environment variable "LLVM_CACHE_REMOTE_SERVICE_SOCKET_PATH".

This requires no linker change to function but ld64 might emit false warnings when using URL schema.

rdar://100281565

jansvoboda11 commented on llvm/include/llvm/CAS/ObjectStore.h:
```diff
@@ -305,8 +305,8 @@ class ObjectProxy {
-  getMemoryBuffer(StringRef Name = "",
-                  bool RequiresNullTerminator = true) const;
+  getMemoryBuffer(StringRef Name = "", bool RequiresNullTerminator = true,
+                  bool Indenpendent = false) const;
```
Typo `Indenpendent` -> `Independent`. Same for the last thunk in `ObjectStore.cpp`.

jansvoboda11 replied:
`Name` and `RequiresNullTerminator ` are fairly self-explanatory, but I'm not sure `Independent` is. Can you document what it does?

cachemeifyoucan replied:
That means the life time is independent from CAS. But just realize that the other function also does the same thing. Remove the unused version.

benlangmuir commented on llvm/include/llvm/LTO/legacy/ThinLTOCodeGenerator.h:
```diff
@@ -125,11 +148,50 @@ class ThinLTOCodeGenerator {
+    StringRef SocketPath = Path;
+    // The environment overwrites the option parameter.
+    if (auto EnvCachePath =
+            sys::Process::GetEnv("LLVM_CACHE_REMOTE_SERVICE_SOCKET_PATH")) {
```
I would prefer we move this logic out of the C++ library code, e.g. to `thinlto_codegen_set_cache_dir` library function and llvm-lto tool.

benlangmuir reviewed (commented):
Is there a reason this is only for remote cas and not a local one?

cachemeifyoucan committed:
[RemoteService] Async pending counter needs to be atomic

NumPending request counter needs to atomic so it will behave correctly
when sending and receiving from different threads.

cachemeifyoucan committed:
[CAS] Remove unused function

Both getMemoryBuffer and loadIndependentDataBuffer return an life time
independent buffer for the CAS data with slightly different
implementation. Remove unused one.

cachemeifyoucan committed:
[ThinLTO][Cache] Allow ThinLTO to use CAS or RemoteService as cache

Allow ThinLTO to use CAS or RemoteService as cache. To use remote
service, set the cache directory to "grpc:/$SERVICE_SOCKET_PATH" or use
OnDisk CAS, set cache directory to "cas:/$CAS_PATH".

rdar://100281565

cachemeifyoucan commented:
Calling RemoveService directly. There seems to be some race condition in my GRPC relay CAS that I haven't had time to figure out. File a separate bug report and convert the CAS path to use on disk CAS since I implemented it already.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan committed:
[RemoteService] Fix llvm-config by making RemoteService a LLVM component

cachemeifyoucan committed:
[RemoteService] Add sync API implementation

ThinLTO and RelayCAS needs sync API. Use an atomic spin lock to block
the async queue to only allow one request.

benlangmuir commented on llvm/include/llvm/LTO/legacy/ThinLTOCodeGenerator.h:
```diff
@@ -125,11 +154,61 @@ class ThinLTOCodeGenerator {
+  Error setCacheDir(std::string Path) {
+    // CacheDir can only be set once.
+    if (!CacheOptions.Path.empty())
+      return Error::success();
```
Curious why this isn't an error?

cachemeifyoucan replied:
The environment variable overwrite happens when codegen is created so it would work even the user didn't specify the cache path through C API. If user did set the cache, it is going to hit the condition and skipped.

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -87,12 +88,19 @@ class KeyValueDBClient {
       --NumPending;
       return receiveNextImpl();
     }
+    Expected<Response> getValueSync(std::string Key) {
```
The new `*Sync` functions seem out-of-place when put inside the `*AsyncQueue` classes, how about moving them to the top classes, `KeyValueDBClient` and `CASDBClient`?

cachemeifyoucan replied:
The stub is owned by queue currently but we can change that for sure. There should be more code sharing as well, which I am planning to improve. Let me give it another thought.

akyrtzi replied:
AFAICT the stub is owned by `CASDBClientImpl` and `KeyValueDBClientImpl`, the queues get references to the stub.

akyrtzi commented:
Could you double-check that when multiple threads call the sync APIs that the gRPC requests go through to the remote service concurrently, instead of the threads waiting for the requests to get responded one-at-a-time?

For example, you could add some print statements to `llvm-remote-cache-service` and use it to verify that it receives multiple concurrent requests while doing thinLTO caching.

cachemeifyoucan commented:
> Could you double-check that when multiple threads call the sync APIs that the gRPC requests go through to the remote service concurrently, instead of the threads waiting for the requests to get responded one-at-a-time?

I am pretty sure the sync APIs are blocking but I don't know if it is blocking on the server side or client side in the queue. Use Sync API only to avoid major re-architecture current logic since it is going to be a lot slower.

akyrtzi commented:
> > Could you double-check that when multiple threads call the sync APIs that the gRPC requests go through to the remote service concurrently, instead of the threads waiting for the requests to get responded one-at-a-time?
> 
> I am pretty sure the sync APIs are blocking but I don't know if it is blocking on the server side or client side in the queue. Use Sync API only to avoid major re-architecture current logic since it is going to be a lot slower.

Using sync API is the right approach, but we should ensure the requests go to the service concurrently. If they don't now we could try creating multiple stubs, one for each thread, or multiple channels, or multiple `CASDBClient`s, etc.

cachemeifyoucan committed:
[GRPCClient] Fix Sync/Async laytout

Move Sync APIs out of AsyncQueues.

cachemeifyoucan commented:
Fix the API layout. Unfortunately server stalls when using syncAPI, which doesn't matter much for using a local CAS but can be very slow for remote CAS. I don't think multi stub is a good idea (that means we need to load balancing from client somehow). I have some idea to rewrite thinLTO codegen into async APIs so I will leave that for later improvement. Now just use SyncAPI, and sync api also makes GRPC relay CAS thread safe.

@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented:
> Unfortunately server stalls when using syncAPI, which doesn't matter much for using a local CAS but can be very slow for remote CAS.

It's a bit unclear to me, could you clarify what you mean with "server stalls"?

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -61,6 +62,35 @@ class KeyValueDBClient {
+  struct PutValueResType {
+    std::shared_ptr<AsyncCallerContext> CallCtx;
+  };
+  Expected<GetValueResType> getValueSync(std::string Key) {
```
It'd be better to separate the return values of the sync APIs from the async ones. `std::shared_ptr<AsyncCallerContext> CallCtx` has no benefit being in the return structure of a sync API, you can simplify the sync APIs and make them more convenient to use by returning the resulting values directly; for example:
```
Expected<Optional<ValueTy>> getValueSync(std::string Key);
Error putValueSync(std::string Key, const ValueTy &Value)
```

cachemeifyoucan commented:
> > Unfortunately server stalls when using syncAPI, which doesn't matter much for using a local CAS but can be very slow for remote CAS.
> 
> It's a bit unclear to me, could you clarify what you mean with "server stalls"?

It means server can only see and serve one request at a time, so it serializes all the requests and it is very slow when you need to download the artifact one by one.

akyrtzi commented:
> It means server can only see and serve one request at a time, so it serializes all the requests and it is very slow when you need to download the artifact one by one.

Does this mean the sync APIs can only serve one thread at a time? Could you try creating one-off `KeyValueDBClients` and `CASDBClients` to serve the LTO requests as they come in, instead of creating one instance that is shared across threads?

cachemeifyoucan commented:
> Does this mean the sync APIs can only serve one thread at a time? Could you try creating one-off KeyValueDBClients and CASDBClients to serve the LTO requests as they come in, instead of creating one instance that is shared across threads?

Sure I can do the experiment. My current understanding is that the un-exposed server completion queue is blocking when sync API is used so that might not help.

cachemeifyoucan committed:
[ThinLTO][CAS] Add Debug options

cachemeifyoucan committed:
[CAS][ThinLTO] Use file based API to optimize transfers

Use file system to transfer LTO generated objects to better optimize for
speed and not hitting the limit of GRPC size limit.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan committed:
[ThinLTO] Add an environmental variable to enable remote CAS

Add an additional environmental variable to control whether we use
the remote caching service provided in the environment.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@akyrtzi @benlangmuir Any more comments on this version?

akyrtzi commented:
Is it making concurrent load/save requests to the service?

benlangmuir commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -89,6 +94,32 @@ namespace {
 static cl::opt<int> ThreadCount("threads", cl::init(0));
+static cl::opt<bool> CacheLogging(
+    "thinlto-cache-logging", cl::desc("Enable logging for thinLTO caching"),
+    cl::init((bool)sys::Process::GetEnv("LLVM_THINLTO_CACHE_LOGGING")),
```
Can we avoid getenv here in the C++ implementation and move it up to liblto and/or whatever executable calls this? Same for the other one below. This pattern hides impure checks deep in the library code.

cachemeifyoucan replied:
I don't see this make any difference. Also the environment checks doesn't affect the output of the computation.

benlangmuir replied:
If we capture stderr as part of the output of the linker like we do in clang, this affects the output since you're logging to stderr.  It looks like the experimental lld caching does capture stderr, for example. I know that doesn't apply to thinlto caching that you are adding here, but it seems like a bad idea to have getenv calls here.  

I agree the determinism check is technically fine, though personally I would prefer we have no getenv at all in the underlying implementation.

cachemeifyoucan replied:
LLD cannot hit this code path. Everything in this file is libLTO only. The logging is for service, we do not replay the log. This is the same as the remarks for cache hit/miss, it is going to be different for replay.

cachemeifyoucan commented:
> Is it making concurrent load/save requests to the service?

Yes. There is potential to add even more concurrency but I will leave that for future.

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -155,6 +174,51 @@ class CASDBClient {
 public:
   virtual ~CASDBClient() = default;
 
+  struct LoadRespond {
```
`LoadRespond` -> `LoadResponse`?

akyrtzi commented on llvm/include/llvm/RemoteCachingService/Client.h:
```diff
@@ -155,6 +174,51 @@ class CASDBClient {
+    bool KeyNotFound = false;
+    Optional<std::string> BlobData;
+  };
+  struct GetRespond {
```
`GetRespond` -> `GetResponse`?

akyrtzi commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -439,6 +452,259 @@ class ModuleCacheEntry {
+    ProducedOutput = true;
+
+    SmallString<64> ResultPath;
+    Expected<sys::fs::file_t> FDOrErr = sys::fs::openNativeFileForRead(
```
Could you add a comment to clarify why you need to "openNativeFileForRead / getOpenFile / closeFile" instead of just calling "MemoryBuffer::getFile"? It's not clear to me.
(Same comment for the other occurrence of this).

cachemeifyoucan replied:
Ah, this is copy-paste from some old code in ThinLTO caching that does buffer reloading from files. I will switch to MemoryBuffer::getFile.

akyrtzi commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -439,6 +452,259 @@ class ModuleCacheEntry {
+    ProducedOutput = true;
+    auto SaveResponse = Service.CASDB->saveFileSync(OutputPath);
+    if (!SaveResponse) {
+      consumeError(SaveResponse.takeError());
```
I'd prefer not to hide error responses, is it possible to report a failure if this occurs? (same comment for the other places that do `consumeError`).

cachemeifyoucan replied:
Just print something to stderr or error out? While none of those are expected to fail but the recovering from thinLTO pipeline is super easy and user will not be hit with error when service is not available.

akyrtzi replied:
I'd like to error out. If I expect to be using the service during building, but the service is have trouble for some reason, I'd like to get the failure as an error.

Whether the service wants to provide an error or not it's up to the service (e.g. maybe the service chooses to ignore network errors), but if the service provides an error we should fail the build and report it.

akyrtzi commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -971,9 +1302,24 @@ void ThinLTOCodeGenerator::run() {
-  if (SavedObjectsDirectoryPath.empty())
+
+  // Need SavedObjectsDirectory for remote cache.
+  bool UseBufferAPI = SavedObjectsDirectoryPath.empty();
```
Shouldn't `bool UseBufferAPI = SavedObjectsDirectoryPath.empty();` move after the following code that initializes `SavedObjectsDirectoryPath`?

cachemeifyoucan replied:
This is intended behavior. We need to force a `SavedObjectDirectory` when we use remote CAS as the place where the service pass file back, but we need to respect the original setting for whether pass the file or buffer back to the caller (indicated by whether there is an original setting on the object directory). I will add some comments to the code.

akyrtzi replied:
To make sure I understand, the linker uses `SavedObjectsDirectoryPath` to indicate whether it wants to receive objects as files or as buffers?

cachemeifyoucan replied:
correct.

akyrtzi commented on llvm/lib/RemoteCachingService/Client/Client.cpp:
```diff
@@ -38,6 +38,119 @@ static Error errorFromGRPCStatus(const grpc::Status &Status) {
 }
 
+static Expected<Optional<KeyValueDBClient::ValueTy>>
+createGetValueRespond(const GetValueResponse &Resp) {
```
`createGetValueRespond` -> `createGetValueResponse`? Same for the other functions below.

akyrtzi commented on llvm/test/CAS/thinlto-remote-cache.ll:
```diff
@@ -0,0 +1,42 @@
+; RUN: llvm-remote-cache-test -socket-path=%{remote-cache-dir}/%basename_t -cache-path=%t/cache -- \
+; RUN:    llvm-lto -thinlto-action=run -exported-symbol=globalfunc %t2.bc %t.bc -thinlto-cache-dir grpc:%{remote-cache-dir}/%basename_t -thinlto-save-objects %t/objects
+; RUN: llvm-cas --cas %t/cache/cas --dump | FileCheck %s
+
```
Should we also have a test invocation that uses the environment variable for setting up the remote cache?

cachemeifyoucan replied:
Not with the current implementation as the environment variable overwrite is only in libLTO, so it requires `ld64` to test. I am not too worried about that.

cachemeifyoucan committed:
Address review feedback

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -439,6 +445,240 @@ class ModuleCacheEntry {
+
+    auto Proxy = CAS.createProxy(None, OutputBuffer.getBuffer());
+    if (!Proxy) {
+      // Ignore error and return.
```
For consistency, could you also report the `LLVMCAS` errors?

cachemeifyoucan commented:
@swift-ci please test

akyrtzi commented on llvm/lib/LTO/ThinLTOCodeGenerator.cpp:
```diff
@@ -439,6 +445,233 @@ class ModuleCacheEntry {
+      report_fatal_error(Proxy.takeError());
+
+    auto Err = Cache.put(*ID, Proxy->getID());
+    report_fatal_error(std::move(Err));
```
Need to check for error before calling `report_fatal_error`, right? 🙂

cachemeifyoucan replied:
Nice catch! Fixed.

cachemeifyoucan commented:
@swift-ci please test
